Document structured logging contribution process (#5812)
kubernetes/community: contributors/devel/sig-instrumentation/migration-to-structured-logging.md

[Structured Logging KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging

> ## Information for Contributors

Reviewer: ```suggestion
## How to contribute
```

> ### How migration is done 

Reviewer: ```suggestion
### About the migration
```

> We organize migration component by component with time in between to collect feedback. 
> Migration of components that are not marked for it is on best effort and is expected to get delayed review or be rejected (example kubeadm).

Reviewer: ```suggestion
Migration proceeds in two different ways. First is organized milestone based large scale migration where we are trying to achieve a particular target for a Kubernetes release (for example migrate whole Kubelet code in 1.21). Second is not organized migration done by individual contributors.

Organized migration is done in two state cycle aligned with Kubernetes releases. In stage one we pick a particular migration milestone and create issue that splits the work into smaller chunks (for example migrating just single file) that can be easily picked up by contributors. This way we ensure that multiple people can independently work on common goal and avoid conflicts. After migration is done we can utilize tooling we developed to mark particular directory as migrated to avoid regressions. When we finish particular milestone we move the second stage where we take a break from migration, to analyse results and improve on the process. Adding structured information to logs is very costly and time consuming process, we don't want to make a mistake that would require us to repeat the migration. That's why we reserve time to collect feedback, analyse impact of changes we have done on log volume, performance and PR review process.  

Non-organized migration is done by individual contributors that want particular code section to utilize new features early. Migration of Kubernetes components not yet marked for migration is on best effort bases. This means that we don't guarantee a timely review or that PR will be accepted at all. It's up to component owners to decide if they want to accept those contributions. As this migration is not organized it means that two contributors could start working on migrating same code. Before sending a PR please make sure that there is was already created by someone else.
```

> ### What's the current status of migration 

Reviewer: ```suggestion
### Current status
```

> Kubelet was migrated, we are collecting feedback

Reviewer: ```suggestion
* 1.21 Kubelet was migrated
* 1.22 We  collecting feedback and making improvements to the migration process.
```

> ### How to get review

Reviewer: It's weird to have "how to get review" section before "create pr requirements" as usually those steps will be done in reverse order. First you create PR then ask for review.

Let's create one section called "Sending a Structured Logging Migration PR"

Author: ok

[Structured Logging KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging

## How to contribute

### About the migration

We organize migration component by component with time in between to collect feedback. 
Migration of components that are not marked for it is on best effort and is expected to get delayed review or be rejected (example kubeadm).

### Current status

* 1.21 Kubelet was migrated

> * 1.22 We  collecting feedback and making improvements to the migration process.

Reviewer: ```suggestion
* 1.22 We are collecting feedback and making improvements to the migration process.
```

Non-organized migration is done by individual contributors that want particular code section to utilize new features early. Migration of Kubernetes components not yet marked for migration is on best effort bases. This means that we don't guarantee a timely review or that PR will be accepted at all. It's up to component owners to decide if they want to accept those contributions. As this migration is not organized it means that two contributors could start working on migrating same code. Before sending a PR please make sure that there is was already created by someone else.

### Current status

* 1.21 Kubelet was migrated
* 1.22 We are collecting feedback and making improvements to the migration process.

> ### Create pr requirements
> 
> * It is recommended to create pr according to a package or feature, not one file
> 
> * Migrate logs according to log content, reject only rename fuction or modify letter case simply
> 
> * Structured Logging migration needs to follow the rules below
> 
> ## Sending a Structured Logging Migration PR
> 
> label with /wg structured-logging and cc @kubernetes/wg-structured-logging-reviews

Reviewer: ```suggestion
## Sending a Structured Logging Migration Pull Request

### Before creating the Pull Request

* Read and understood the. In case of any questions, please contact us on [#wg-structured-logging](https://kubernetes.slack.com/messages/wg-structured-logging) Slack channel.
* Check list of [opened PRs](https://github.com/kubernetes/kubernetes/pulls?q=is%3Aopen+label%3Awg%2Fstructured-logging+is%3Apr) to check if someone is already working on this.

### Getting review

To make sure PR is reviewed in timely manner please add "wg-structured-logging" label and notify our reviewers. To do that please run commends below:

/wg structured-logging 
/cc @kubernetes/wg-structured-logging-reviews

### Why my PR was rejected?

Fact that Kubernetes is organizing migration of to Structured Logging doesn't mean that are accepting all the PRs that migrate logs. We reserve the right to reject the Pull Request in situations listed below:

* Pull request is below minimum quality standards and clearly shows that author hasn't read the guide at all, for example just renamed 'Infof" to 'InfoS'.
* Pull request migrates components which owners decided to not conduct migration at this time. List of those components: kubeadm.

 
```

Non-organized migration is done by individual contributors that want particular code section to utilize new features early. Migration of Kubernetes components not yet marked for migration is on best effort bases. This means that we don't guarantee a timely review or that PR will be accepted at all. It's up to component owners to decide if they want to accept those contributions. As this migration is not organized it means that two contributors could start working on migrating same code. Before sending a PR please make sure that there is was already created by someone else.

### Current status

* 1.21 Kubelet was migrated
* 1.22 We are collecting feedback and making improvements to the migration process.

## Sending a Structured Logging Migration Pull Request

### Before creating the Pull Request

* Read and understood the. In case of any questions, please contact us on [#wg-structured-logging](https://kubernetes.slack.com/messages/wg-structured-logging) Slack channel.
* Check list of [opened PRs](https://github.com/kubernetes/kubernetes/pulls?q=is%3Aopen+label%3Awg%2Fstructured-logging+is%3Apr) to check if someone is already working on this.

> ### Getting review
> 
> To make sure PR is reviewed in timely manner please add "wg-structured-logging" label and notify our reviewers. To do that please run commends below:
> 
> /wg structured-logging 
> /cc @kubernetes/wg-structured-logging-reviews

Reviewer: ```suggestion
### What to include in the Pull request

To make sure PR is reviewed in timely manner we require authors to provide necessary information for it to be properly triaged and reviewed by knowledgeable parties.

* title: `Migrate <directory/file> to structured logging`
* WGs: `structured-logging`
* area: `logging`
* priority: `important-longterm` for normal PR or `important-soon` for PRs that are part of organized migration of component. 
* cc: @kubernetes/wg-structured-logging-reviews
* kind: cleanup
* release note: `Migrate <directory/file> to structured logging
`
To quickly add this information you can comment on PR with below commands:

/retitle Migrate <directory/file> to structured logging
/wg structured-logging 
/area logging
/priority important-longterm
/king cleanup
/cc @kubernetes/wg-structured-logging-reviews

And edit the top comment to include release note:
Migrate <directory/file> to structured logging
```

[Structured Logging KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging

## How to contribute

### About the migration

> Migration proceeds in two different ways. First is organized milestone based large scale migration where we are trying to achieve a particular target for a Kubernetes release (for example migrate whole Kubelet code in 1.21). Second is not organized migration done by individual contributors.

Reviewer 2: Reworded as -

Migration within the structured logging working group happens in two different ways - organized & non-organized. With organized, milestone-based, large-scale migration we try to target deliverables for a specific Kubernetes release. A good example of such an effort is the migration of the entire Kubelet code in the 1.21 release <link_to_the_docs>. For non-organized migrations, the onus is, generally, on the individual contributors.

> 

Reviewer 2: #### Organized Migration

> Organized migration is done in two state cycle aligned with Kubernetes releases. In stage one we pick a particular migration milestone and create issue that splits the work into smaller chunks (for example migrating just single file) that can be easily picked up by contributors. This way we ensure that multiple people can independently work on common goal and avoid conflicts. After migration is done we can utilize tooling we developed to mark particular directory as migrated to avoid regressions. When we finish particular milestone we move the second stage where we take a break from migration, to analyse results and improve on the process. Adding structured information to logs is very costly and time consuming process, we don't want to make a mistake that would require us to repeat the migration. That's why we reserve time to collect feedback, analyse impact of changes we have done on log volume, performance and PR review process.  

Reviewer 3: qq - what tooling are you using? 
it might be worth linking out to it so folks can take a look at it.

Reviewer: We added a a verify script https://github.com/kubernetes/kubernetes/blob/master/hack/verify-structured-logging.sh
It checks directories listed in https://github.com/kubernetes/kubernetes/blob/master/hack/.structured_logging using a tool we created in https://github.com/kubernetes/klog/tree/main/hack/tools/logcheck.

> Organized migration is done in two state cycle aligned with Kubernetes releases. In stage one we pick a particular migration milestone and create issue that splits the work into smaller chunks (for example migrating just single file) that can be easily picked up by contributors. This way we ensure that multiple people can independently work on common goal and avoid conflicts. After migration is done we can utilize tooling we developed to mark particular directory as migrated to avoid regressions. When we finish particular milestone we move the second stage where we take a break from migration, to analyse results and improve on the process. Adding structured information to logs is very costly and time consuming process, we don't want to make a mistake that would require us to repeat the migration. That's why we reserve time to collect feedback, analyse impact of changes we have done on log volume, performance and PR review process.  

Reviewer 2: Reworded as,

Organized migration is carried out in a two state cycle aligning with the cadence of Kubernetes releases. In the first stage, we pick a particular migration milestone & create an issue to split the work into smaller chunks (for example, migrating just a single file). This ensures that the work can be easily picked up by multiple individual contributors and also avoids conflicts. After the migration activity is complete, we mark the directory as migrated to avoid regressions with the help of tooling that we have developed. In the second milestone, we take a break from migration to analyze results & improve our existing processes. Adding structured information to logs is a very costly & time-consuming affair. Setting aside time in the second stage to collect feedback, analyze the impact of changes on the log volume & performance, and better our PR review process helps us avoid mistakes and duplicate efforts.

> 

Reviewer 2: #### Non-organized migration

> Non-organized migration is done by individual contributors that want particular code section to utilize new features early. Migration of Kubernetes components not yet marked for migration is on best effort bases. This means that we don't guarantee a timely review or that PR will be accepted at all. It's up to component owners to decide if they want to accept those contributions. As this migration is not organized it means that two contributors could start working on migrating same code. Before sending a PR please make sure that there is was already created by someone else.

Reviewer 2: As aforementioned, our non-organized migration efforts are spearheaded by individual contributors who require particular code sections to utilize new features early. Efforts for Kubernetes components not yet marked for migration also fall under this category & are carried out on a best-efforts basis. This means that we don't guarantee a timely review or that the PR will be accepted at all. The respective component owners have a final say in the acceptance of these contributions. Since this is a non-organized effort there is a high possibility that two contributors could be working on migrating the same code. Before sending a PR our way, please ensure that there isn't one already in place created by someone else.

### Current status

* 1.21 Kubelet was migrated
* 1.22 We are collecting feedback and making improvements to the migration process.

## Sending a Structured Logging Migration Pull Request

### Before creating the Pull Request

> * Read and understood the. In case of any questions, please contact us on [#wg-structured-logging](https://kubernetes.slack.com/messages/wg-structured-logging) Slack channel.

Reviewer 2: "Read and understood the" -> Incomplete sentence.

> * Check list of [opened PRs](https://github.com/kubernetes/kubernetes/pulls?q=is%3Aopen+label%3Awg%2Fstructured-logging+is%3Apr) to check if someone is already working on this.

Reviewer 2: Check list of [opened PRs](https://github.com/kubernetes/kubernetes/pulls?q=is%3Aopen+label%3Awg%2Fstructured-logging+is%3Apr) to understand if someone is already working on this.

### What to include in the Pull request

> To make sure PR is reviewed in timely manner we require authors to provide necessary information for it to be properly triaged and reviewed by knowledgeable parties.

Reviewer 2: Reworded:

To ensure that the PRs are reviewed in a timely manner, we require authors to provide the below information. This helps in proper triaging & reviewing by the concerned knowledgeable parties.

* title: `Migrate <directory/file> to structured logging`
* WGs: `structured-logging`
* area: `logging`
* priority: `important-longterm` for normal PR or `important-soon` for PRs that are part of organized migration of component. 
* cc: @kubernetes/wg-structured-logging-reviews
* kind: cleanup
* release note: `Migrate <directory/file> to structured logging
`
To quickly add this information you can comment on PR with below commands:

/retitle Migrate <directory/file> to structured logging
/wg structured-logging 
/area logging
/priority important-longterm
/king cleanup

> /cc @kubernetes/wg-structured-logging-reviews

Reviewer: Can you surround this into triple quotes back-ticks ` ``` `

Author: done

And edit the top comment to include release note:
Migrate <directory/file> to structured logging

### Why my PR was rejected?

> Fact that Kubernetes is organizing migration of to Structured Logging doesn't mean that are accepting all the PRs that migrate logs. We reserve the right to reject the Pull Request in situations listed below:

Reviewer 2: Reworded:

Even though the Kubernetes project is organizing migration of Structured Logging, this doesn't mean that we are able to accept all the PRs that come our way. We reserve the right to reject the Pull Request in situations listed below:

> * Pull request is below minimum quality standards and clearly shows that author hasn't read the guide at all, for example just renamed 'Infof" to 'InfoS'.

Reviewer: ```suggestion
* Pull request is below minimum quality standards and clearly shows that author hasn't read the guide at all, for example PR just renames 'Infof" to 'InfoS'.
```

> * Pull request migrates components which owners decided to not conduct migration at this time. List of those components: kubeadm.

Reviewer 2: Reworded:

* Pull request migrates components that the owners have decided against migrating. List of those components: kubeadm.

* title: `Migrate <directory/file> to structured logging`
* WGs: `structured-logging`
* area: `logging`
* priority: `important-longterm` for normal PR or `important-soon` for PRs that are part of organized migration of component. 
* cc: @kubernetes/wg-structured-logging-reviews
* kind: cleanup
* release note: `Migrate <directory/file> to structured logging
`
To quickly add this information you can comment on PR with below commands:

```
/retitle Migrate <directory/file> to structured logging
/wg structured-logging 
/area logging
/priority important-longterm
/king cleanup
/cc @kubernetes/wg-structured-logging-reviews
```
And edit the top comment to include release note:
Migrate <directory/file> to structured logging

### Why my PR was rejected?

Even though the Kubernetes project is organizing migration of Structured Logging, this doesn't mean that we are able to accept all the PRs that come our way. We reserve the right to reject the Pull Request in situations listed below:

* Pull request is below minimum quality standards and clearly shows that author hasn't read the guide at all, for example PR just renames 'Infof" to 'InfoS'.

> * Pull request migrates components that the owners have decided against migrating. List of those components: kubeadm.

Reviewer 4: ```suggestion
* Pull request migrates components that the owners have decided against migrating. List of those components:
   * kubeadm
```

For non-organized migrations, the onus is, generally, on the individual contributors.

* Organized Migration

Organized migration is carried out in a two state cycle aligning with the cadence of Kubernetes releases. 

In the first stage, we pick a particular migration milestone & create an issue to split the work into smaller chunks (for example, migrating just a single file). This ensures that the work can be easily picked up by multiple individual contributors and also avoids conflicts. After the migration activity is complete, we mark the directory as migrated to avoid regressions with the help of tooling that we have developed. 

In the second milestone, we take a break from migration to analyze results & improve our existing processes. Adding structured information to logs is a very costly & time-consuming affair. Setting aside time in the second stage to collect feedback, analyze the impact of changes on the log volume & performance, and better our PR review process helps us avoid mistakes and duplicate efforts.

* Non-organized migration

> As aforementioned, our non-organized migration efforts are spearheaded by individual contributors who require particular code sections to utilize new features early. 

Reviewer 5: Not quite sure what's meant by this - can you elaborate?

Author: I think the meaning here is that
can migrate separately when there is no migirage plan for the particular code sections but want to utilize new features early.

Reviewer 5: I'd re-word this as follows:
```suggestion
As aforementioned, our non-organized migration efforts are spearheaded by individual contributors who need to migrate particular code sections to utilize new features early. 
```

> Efforts for Kubernetes components not yet marked for migration also fall under this category & are carried out on a best-efforts basis. This means that we don't guarantee a timely review or that the PR will be accepted at all. 

Reviewer 5: I'd re-word this to be more welcoming: we will try our best to review and accept as many PRs as we can, but being a small team, we can't give any time frames for the same.

Author: done, Thank you very much.

The respective component owners have a final say in the acceptance of these contributions. Since this is a non-organized effort there is a high possibility that two contributors could be working on migrating the same code. 

Before sending a PR our way, please ensure that there isn't one already in place created by someone else.

### Current status

* 1.21 Kubelet was migrated
* 1.22 We are collecting feedback and making improvements to the migration process.

## Sending a Structured Logging Migration Pull Request

### Before creating the Pull Request

> * Incomplete sentence. In case of any questions, please contact us on [#wg-structured-logging](https://kubernetes.slack.com/messages/wg-structured-logging) Slack channel.

Reviewer 5: What's meant by `Incomplete Sentence` here?

* Check list of [opened PRs](https://github.com/kubernetes/kubernetes/pulls?q=is%3Aopen+label%3Awg%2Fstructured-logging+is%3Apr) to understand if someone is already working on this.

### What to include in the Pull request

To ensure that the PRs are reviewed in a timely manner, we require authors to provide the below information. This helps in proper triaging & reviewing by the concerned knowledgeable parties.

* title: `Migrate <directory/file> to structured logging`
* WGs: `structured-logging`
* area: `logging`
* priority: `important-longterm` for normal PR or `important-soon` for PRs that are part of organized migration of component. 
* cc: @kubernetes/wg-structured-logging-reviews
* kind: cleanup
* release note: `Migrate <directory/file> to structured logging
`
To quickly add this information you can comment on PR with below commands:

```
/retitle Migrate <directory/file> to structured logging
/wg structured-logging 
/area logging
/priority important-longterm

> /king cleanup

Reviewer 5: ```suggestion
/kind cleanup
```

/cc @kubernetes/wg-structured-logging-reviews
```
And edit the top comment to include release note:
Migrate <directory/file> to structured logging

### Why my PR was rejected?

Even though the Kubernetes project is organizing migration of Structured Logging, this doesn't mean that we are able to accept all the PRs that come our way. We reserve the right to reject the Pull Request in situations listed below:

> 

Reviewer 5: I'd re-word this be more welcoming and friendly like the example I gave above

> * Pull request is below minimum quality standards and clearly shows that author hasn't read the guide at all, for example PR just renames 'Infof" to 'InfoS'.

Reviewer 4: ```suggestion
* Pull request is below minimum quality standards and clearly shows that author hasn't read the guide at all, for example PR just renames `Infof` to `InfoS`.
```
